src/app/form-password/form-password.component.ts

```diff
@@ -0,0 +1,47 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable no-unused-expressions */
```

Comment:
usare SOLO eslint-disable-next-line, MAI eslint-disable sull'intero file

Reply:
Cristo, allora mi tocca metterlo 16 volte tipo ahah

Reply:
in realtà non dovresti proprio metterlo :D 90% dei casi il warning è indicativo di un problema. tipo tutti gli any che hai usato sono mega malvagi perché è come se in java usassi object per tutte le cose che passi

Reply:
si però dobbiamo approfondire un attimo meglio questa cosa, perchè alcune delle funzioni dei loro moduli hanno come type any e quello che mi ritornano a volte cambia a seconda di cosa succede in quella chiamata

Reply:
typescript è molto flessibile al riguardo nel senso che i type possono venir definiti con l'OR:

https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions

chiaramente di quel tutorial non si utilizza lo switch ma una factory o semplicemente una type guard con un if:

https://basarat.gitbook.io/typescript/type-system/typeguard#literal-type-guard

un esempio con una roba nullabile:

```ts
type Optional = string | null

// getValueFromSomewhere non dichiara un tipo ma any e noi sappiamo che può essere una string o un null
const value = getValueFromSomewhere()

if (!value) {
 throw "i got null!"
}

// ... (1)
```

da (1) in poi typescript è abbastanza furbo da inferire automaticamente che il tipo Optional è "collassato" a string.

Chiaramente fare ste cose comporta fatica ma i benefici sono ENORMI in termini di qualità del codice

Reply:
btw le due cose che mi piacciono di più di quello che facciamo sono

1. configurare roba
2. tipizzare roba

Reply:
> quello che mi ritornano a volte cambia a seconda di cosa succede in quella chiamata

il vantaggio di fare così è proprio che stiamo facendo leva su funzionalità intrinseche del linguaggio per effettuare validazione su quello che è successo. Se non l'avessi gestito così avresti dovuto schiaffarci dentro un sacco di condizionali per discriminare le azioni da compiere; facendolo _direttamente_ in questo modo otteniamo lo stesso risultato ma è il compilatore stesso che ci dice cos'è andato storto. E in termini di performance è risolto TUTTO a compile time quindi l'impatto è nullo!